MaxText/layers/moe.py

```diff
@@ -357,29 +361,115 @@ def unpermute(self, intermediate, sorted_selected_experts, weights, batch_size,
       )
     return output.reshape(batch_size, sequence_length, -1).astype(self.dtype)
 
-  def sparse_matmul(self, inputs, gate_logits, pre_bias_logits, w0_kernel, w1_kernel, wo_kernel):
-    """sparse matrix multiplication"""
-    tile_size = (512, 1024, 1024)  # (m, k, n)
-
-    def local_permute(inputs, global_group_sizes, local_expert_size):
-      """Sort inputs by expert within each shard."""
-      local_id = jax.lax.axis_index("expert")
-      local_sizes = jax.lax.dynamic_slice_in_dim(
-          global_group_sizes, local_id * local_expert_size, local_expert_size, axis=1
-      ).reshape(-1)
+
+  @staticmethod
+  def local_permute(inputs, global_group_sizes, local_expert_size,
+                    shard_index, is_offset=False, global_sorted_experts=None):
+    """Sort inputs by expert within each shard."""
+    all_shard_local_sizes = jax.lax.dynamic_slice_in_dim(
+        global_group_sizes, shard_index * local_expert_size, local_expert_size, axis=1
+    )
+    local_sizes = all_shard_local_sizes.reshape(-1)
+    local_group_size = jnp.sum(all_shard_local_sizes, axis=0)
+    
+    # In this case, the data that needs to be processed by the local shard
+    # does not start from row 0 but actually starts at
+    # jnp.concatenate((jnp.array([0]), jnp.cumsum(local_group_sizes[:-1]))[shard_id]
+    # This happens for instance if the data is not produced via a ragged_all_to_all.
+    if is_offset:
+      divided_assignments = jnp.floor_divide(global_sorted_experts, local_expert_size)
+      expert_indices = jnp.where(
+        divided_assignments == shard_index,
+        jnp.mod(global_sorted_experts, local_expert_size),
+        local_expert_size
+      )
+    else:
       base_indices = jnp.mod(jnp.arange(local_sizes.shape[0]), local_expert_size)
       expert_indices = jnp.repeat(base_indices, local_sizes, total_repeat_length=inputs.shape[0])
-      sorted_indices = jnp.argsort(expert_indices)
-      sorted_inputs = jnp.take(inputs, indices=sorted_indices, axis=0)
-      group_size = jnp.bincount(expert_indices, length=local_expert_size)
-      return sorted_inputs, sorted_indices, group_size
+    sorted_indices = jnp.argsort(expert_indices)
+    sorted_inputs = jnp.take(inputs, indices=sorted_indices, axis=0)
+    sorted_experts_ids = expert_indices[sorted_indices]
```

Comment:
Thanks for the change! I will pull down and see if this improves the training side.